src/sources/host_metrics.rs

```diff
@@ -545,13 +545,10 @@ impl HostMetricsConfig {
                     .filter_map(|partition| async { partition })
                     // Filter on configured devices
                     .map(|partition| {
-                        (self.filesystem.devices.is_empty()
-                            && partition
```

Comment:
I could have also switched this to `||` here, but removing that part of the boolean check seems to match up better with how the other filtering is done.

Reply:
I think you're right on both counts. The `&&` is definitely wrong here. I broke the logic when I changed to this from an `if/else` block. Something is nagging me that the the `is_empty` guard is still needed (with a `||` of course) due to the desired semantics for when the filesystem has no device, but I can't work out when that might be the case.

Reply:
If the filesystem has no device it seems like the `map()` below would just skip any filtering and then it would default to returning `true` via the `unwrap_or()`.

I think this is OK behavior for this circumstance though I could maybe see users wanting to exclude filesystems with no devices if they specify an `includes` list.

Reply:
Ah yes, that was it. The `includes` list has the semantics of "only include the ones listed here". However, the `unwrap` would *also* include all the no-device filesystems, which is not desired.

Reply:
I think the `if_empty` guard won't result in the right behavior either though.

Imagine I specify `includes = ["foo"]`. `if_empty()` would return `true` and then it would do the below check by `map`ing and return the default of `true` if the filesystem has no device.

I think we want `unwrap_or(false)` with the `is_empty` check, but that won't cover the case where only `excludes` are supplied. That is, if I specify only `excludes = ["foo"]` I'd still expect to see filesystems with no devices.

This does highlight the possible need to explicitly include or exclude filesystems with no devices with some sort of "null" value. For example, `includes = ["foo", null]`. Perhaps some sort of sentinel string like `$empty` or `$null` could be used. What do you think?

Reply:
Hmm, yes, this logic defies a simple resolution. It might need a special matcher in `FilterList`.

There would be value in being able to match filesystems with no device, however I would say it is out of the scope of this issue and good for follow-up work.

Reply:
:+1: I pushed https://github.com/timberio/vector/pull/6921/commits/6f2bfa750181c279132384dd34266d1e8d1f2bd4 which rectors the filtering logic to handle when the value being filtered on is not there. If `includes` is specified, and the value is `None`, it is excluded. If `excludes` is specified, and the value is `None`, it is allowed through.

Does that match what you were thinking?

I'll open a separate issue for allowing explicitly including or excluding null values.

Reply:
Follow-up issue: https://github.com/timberio/vector/issues/6923